array.c

```diff
@@ -2981,6 +2981,40 @@ rb_ary_to_ary_m(VALUE ary)
     return ary;
 }
 
+
+/*
+ *  call-seq:
+ *    array.to_proc -> proc
+ *
+ *  Returns a new \Proc object where the first element is the method and the rest are the arguments:
+ *     arr = [:+, 1]
+ *     proc = arr.to_proc
+ *     proc.class # => Proc
+ *     proc.call(2) # => 3
+ *     [1, 2, 3].map(&proc) # => [2, 3, 4]
+ *
+ *  If the +array+ is empty, raises ArgumentError.
+ *  If the first element of the +array+ is not a Symbol or a String, raises TypeError.
+ */
+
+
+static VALUE
+ary_proc_call(RB_BLOCK_CALL_FUNC_ARGLIST(recv, ary))
+{
+    argc = RARRAY_LENINT(ary);
+    rb_check_arity(argc, 1, UNLIMITED_ARGUMENTS);
+    VALUE method = rb_to_symbol(RARRAY_AREF(ary, 0));
+    VALUE rest = rb_ary_subseq(ary, 1, argc);
+    return rb_funcallv(recv, SYM2ID(method), argc - 1, RARRAY_PTR(rest));
```

Comment:
Not only this `rest` may be optimized away and garbage-collected, it is unnecessary.
```suggestion
    return rb_funcallv(recv, SYM2ID(method), argc - 1, RARRAY_PTR(ary) + 1);
```